### PR TITLE
Enable Iceoryx shared memory support in Cyclone DDS

### PR DESCRIPTION
**Pull request by gmartin82**
This PR adds support for Iceoryx shared memory within Cyclone DDS.

Shared memory support is enabled through the optional **dds_shm** feature and currently has the following limitations:

* Not applicable when the --fwd-discovery option is used.
* Only data readers of compatible types are able to gain any benefit from enabling the feature.
* ROS2 types are unable to benefit from shared memory. This is a limitation of Cyclone DDS's RMW implementation and may be solved in future by [this PR](https://github.com/ros2/rmw_cyclonedds/pull/445).

### CODE REVIEW

File: DEFAULT_CONFIG.json5 [hunk 1]
```diff
@@ -27,6 +27,12 @@
       ////
       // localhost_only: true,
 
+      ////
+      //// shm_enabled: If set to true, the DDS implementation will use Iceoryx shared memory.
```
Comment (JEnoch) [inline DEFAULT_CONFIG.json5:31]: Specify that the bridge has to be built with `dds_shm` feature for this configuration to be applied

File: zenoh-plugin-dds/src/dds_mgt.rs [hunk 2]
```diff
@@ -119,6 +211,38 @@ impl Drop for DDSRawSample {
...
+    }
+}
+
+impl From<DDSRawSample> for Value {
+    fn from(buf: DDSRawSample) -> Self {
+        #[cfg(feature = "dds_shm")]
+        {
+            if let Some(iox_chunk) = buf.iox_chunk {
+                let mut zbuf = ZBuf::default();
+                zbuf.push_zslice(ZSlice::from(buf.data_as_slice().to_vec()));
```
Comment (JEnoch) [inline zenoh-plugin-dds/src/dds_mgt.rs:237]: Why to push a slice from data here? Isn't all in `iox_chunk`?
Please add a comment in code to explain, so we don't wonder in the future.

**Review by JEnoch — CHANGES_REQUESTED** [review]

File: zenoh-plugin-dds/src/dds_mgt.rs (hunk 2 above)
Comment (gmartin82) [reply zenoh-plugin-dds/src/dds_mgt.rs:237]: No, when a shared memory buffer is received only the payload is in the `iox_chunk`. The header is still contained within the iovec data. This is why both are pushed into the buffer. I'll add a comment to clarify.

File: README.md [hunk 3]
```diff
@@ -114,6 +114,24 @@ $ cargo build --release -p zenoh-bridge-dds
 The **`zenoh-bridge-dds`** binary will be generated in the `target/release` sub-directory.
 
 
+### Enabling Cyclone DDS Shared Memory Support
+
+Cyclone DDS Shared memory support is provided by the [Iceoryx library](https://iceoryx.io/). Iceoryx introduces additional system requirements which are documented [here](https://iceoryx.io/v2.0.1/getting-started/installation/#dependencies).
+
+To build the zenoh bridge for DDS with support for shared memory the `dds_shm` optional feature must be enabled during the build process as follows:
```
[the remark this answers is not in the document]
Comment (gmartin82) [reply README.md:121]: Requested information added.

File: README.md [hunk 4]
```diff
@@ -114,6 +114,24 @@ $ cargo build --release -p zenoh-bridge-dds
...
+```bash
+$ cargo build --release -p zenoh-plugin-dds --features dds_shm
+```
+
+- standalone executable binary:
+```bash
+$ cargo build --release -p zenoh-bridge-dds --features dds_shm
+```
+
+See [here](https://cyclonedds.io/docs/cyclonedds/latest/shared_memory/shared_memory.html) for more details of shared memory support in Cyclone DDS.
```
[the remark this answers is not in the document]
Comment (gmartin82) [reply README.md:138]: Requested information added.

File: DEFAULT_CONFIG.json5 (hunk 1 above)
Comment (gmartin82) [reply DEFAULT_CONFIG.json5:31]: Requested information added. I've also added the same warning to the Configuration section in the readme.

**Review by JEnoch — APPROVED** [review]

**Pull request merged**